src/ui/VirtualFileRenderer/VirtualFileRenderer.tsx

```diff
@@ -0,0 +1,323 @@
+import { useWindowVirtualizer } from '@tanstack/react-virtual'
+// eslint-disable-next-line no-restricted-imports
+import type { Dictionary } from 'lodash'
+import Highlight, { defaultProps } from 'prism-react-renderer'
+import { useEffect, useLayoutEffect, useRef } from 'react'
+import { useHistory, useLocation } from 'react-router-dom'
+
+import { requestAnimationTimeout } from 'shared/utils/animationFrameUtils'
+import { cn } from 'shared/utils/cn'
+import { prismLanguageMapper } from 'shared/utils/prismLanguageMapper'
+
+import { ColorBar } from './ColorBar'
+
+// copied from prism-react-renderer since they don't export it
+type Token = {
+  types: string[]
+  content: string
+  empty?: boolean
+}
+
+interface CodeBodyProps {
+  tokens: Token[][]
+  getLineProps: Highlight['getLineProps']
+  getTokenProps: Highlight['getTokenProps']
+  coverage: Dictionary<'H' | 'M' | 'P'>
+  codeDisplayOverlayRef: React.RefObject<HTMLDivElement>
+}
+
+const CodeBody = ({
+  tokens,
+  getLineProps,
+  getTokenProps,
+  coverage,
+  codeDisplayOverlayRef,
+}: CodeBodyProps) => {
+  const history = useHistory()
+  const location = useLocation()
+
+  const initialRender = useRef(true)
+  const wrapperRef = useRef<HTMLDivElement>(null)
+
+  // virtualize code lines
+  const virtualizer = useWindowVirtualizer({
+    count: tokens.length,
+    estimateSize: () => 18,
+    overscan: 75,
+    scrollMargin: codeDisplayOverlayRef.current?.offsetTop ?? 0,
+  })
+
+  const div = codeDisplayOverlayRef.current
+  if (div) {
+    // set the parent div height to the total size of the virtualizer
+    div.style.height = `${virtualizer.getTotalSize()}px`
+    div.style.position = 'relative'
+  }
+
+  // TODO - finalize this logic
+  // scroll to line on the initial render
+  useEffect(() => {
+    if (initialRender.current) {
+      initialRender.current = false
+      const index = parseInt(location.hash.slice(2), 10)
```

Comment:
You need to handle negative or bad line links
```suggestion
      const index = parseInt(location.hash.slice(2), 10)
      if(isNaN(index) || index < 0){
        Sentry.captureMessage or throw error
      }
```

Reply:
Should also probably handle line numbers that are too large.

Reply:
Thoughts on handling only these cases: `!isNaN(index) && index > 0 && index < tokens.length` and ignoring the rest? By default the value is `NaN` so it would constantly throw an error.

Reply:
You mean default value is blank, but gets parsed as NaN? I think a blank value is valid, but something like #Lwat that doesnt parse to a lineno should be handled

Reply:
So this is what I came up with instead, handle the "valid" cases and ignore the rest

```tsx
if (isNumber(index) && index > 0 && index < tokens.length) {
  requestAnimationFrame(() => {
    virtualizer.scrollToIndex(index, { align: 'start' })
  })
}
```